kernel/device.c

```diff
@@ -27,6 +28,58 @@ void z_device_state_init(void)
 	}
 }
 
+
+int do_device_init(const struct device *dev)
+{
+	int rc = 0;
+
+	if (dev->ops.init != NULL) {
+		rc = dev->ops.init(dev);
+		/* Mark device initialized. If initialization
+		 * failed, record the error condition.
+		 */
+		if (rc != 0) {
+			if (rc < 0) {
+				rc = -rc;
+			}
+			if (rc > UINT8_MAX) {
+				rc = UINT8_MAX;
+			}
```

Comment:
This might be beyond the scope of the work item, but if `rc` is ever INT_MIN, then -rc is still less than 0. It may be worthwhile to cast `rc` to an `unsigned int` as below to guarantee that we do not get the weird fringe case.

```
   if ((unsigned int)rc > UINT8_MAX) {
      rc = UINT8_MAX;
   }
```

Reply:
yeah, not part of this PR....